group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -5524,5 +5670,2897 @@ private static Record newGroupMetadataRecord(
+            .withRebalanceTimeoutMs(rebalanceTimeoutMs)
+            .withSessionTimeoutMs(sessionTimeoutMs)
+            .build();
+
+        CompletableFuture<JoinGroupResponseData> leaderJoinResponseFuture = new CompletableFuture<>();
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(joinRequest, leaderJoinResponseFuture);
+
+        assertTrue(result.records().isEmpty());
+
+        CompletableFuture<JoinGroupResponseData> followerJoinResponseFuture = new CompletableFuture<>();
+        result = context.sendGenericGroupJoin(
+            joinRequest.setGroupInstanceId(followerInstanceId),
+            followerJoinResponseFuture);
+
+        assertTrue(result.records().isEmpty());
+        assertFalse(leaderJoinResponseFuture.isDone());
+        assertFalse(followerJoinResponseFuture.isDone());
+
+        // The goal for two timer advance is to let first group initial join complete and set newMemberAdded flag to false. Next advance is
+        // to trigger the rebalance as needed for follower delayed join. One large time advance won't help because we could only populate one
+        // delayed join from purgatory and the new delayed op is created at that time and never be triggered.
+        context.sleepAndAssertEmptyResult(context.genericGroupInitialRebalanceDelayMs);
+        context.sleepAndAssertEmptyResult(context.genericGroupInitialRebalanceDelayMs);
+
+        assertTrue(leaderJoinResponseFuture.isDone());
+        assertTrue(followerJoinResponseFuture.isDone());
+        assertEquals(Errors.NONE.code(), leaderJoinResponseFuture.get().errorCode());
+        assertEquals(Errors.NONE.code(), followerJoinResponseFuture.get().errorCode());
+        assertEquals(1, leaderJoinResponseFuture.get().generationId());
+        assertEquals(1, followerJoinResponseFuture.get().generationId());
+        assertEquals(2, group.size());
+        assertEquals(1, group.generationId());
+        assertTrue(group.isInState(COMPLETING_REBALANCE));
+
+        String leaderId = leaderJoinResponseFuture.get().memberId();
+        String followerId = followerJoinResponseFuture.get().memberId();
+        List<SyncGroupRequestAssignment> assignment = new ArrayList<>();
+        assignment.add(new SyncGroupRequestAssignment().setMemberId(leaderId)
+            .setAssignment(new byte[]{1}));
+        assignment.add(new SyncGroupRequestAssignment().setMemberId(followerId)
+            .setAssignment(new byte[]{2}));
+
+        SyncGroupRequestData syncRequest = new SyncGroupRequestBuilder()
+            .withGroupId(groupId)
+            .withGroupInstanceId(leaderInstanceId)
+            .withMemberId(leaderId)
+            .withGenerationId(1)
+            .withAssignment(assignment)
+            .build();
+
+        CompletableFuture<SyncGroupResponseData> leaderSyncResponseFuture = new CompletableFuture<>();
+        result = context.sendGenericGroupSync(
+                syncRequest,
+                leaderSyncResponseFuture);
```

Comment:
I still this code like this one. Couldn't we rely on `SyncResult` to simplify it as well?